chapter_natural-language-processing-pretraining/subword-embedding.ipynb

```diff
@@ -0,0 +1,423 @@
+    "%maven org.slf4j:slf4j-api:1.7.26\n",
+    "%maven org.slf4j:slf4j-simple:1.7.26\n",
+    "    \n",
+    "%maven ai.djl:api:0.11.0-SNAPSHOT\n",
+    "%maven ai.djl:basicdataset:0.11.0-SNAPSHOT\n",
+    "\n",
+    "// See https://github.com/deepjavalibrary/djl/blob/master/mxnet/mxnet-engine/README.md\n",
+    "// MXNet \n",
+    "%maven ai.djl.mxnet:mxnet-engine:0.11.0-SNAPSHOT\n",
+    "%maven ai.djl.mxnet:mxnet-native-auto:1.7.0-backport"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load ../utils/Functions.java"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ai.djl.ndarray.NDManager;\n",
+    "import ai.djl.translate.TranslateException;\n",
+    "import ai.djl.util.Pair;\n",
+    "\n",
+    "import java.io.IOException;\n",
+    "import java.util.*;\n",
+    "import java.util.stream.Stream;"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "NDManager manager = NDManager.newBaseManager(Functions.tryGpu(0));"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "String[] symbols =\n",
+    "        new String[] {\n",
+    "            \"a\", \"b\", \"c\", \"d\", \"e\", \"f\", \"g\", \"h\", \"i\", \"j\", \"k\", \"l\", \"m\", \"n\", \"o\", \"p\",\n",
+    "            \"q\", \"r\", \"s\", \"t\", \"u\", \"v\", \"w\", \"x\", \"y\", \"z\", \"_\", \"[UNK]\"\n",
+    "        };"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 2
+   },
+   "source": [
+    "Since we do not consider symbol pairs that cross boundaries of words,\n",
+    "we only need a dictionary `rawTokenFreqs` that maps words to their frequencies (number of occurrences)\n",
+    "in a dataset.\n",
+    "Note that the special symbol `'_'` is appended to each word so that\n",
+    "we can easily recover a word sequence (e.g., \"a taller man\")\n",
+    "from a sequence of output symbols ( e.g., \"a_ tall er_ man\").\n",
+    "Since we start the merging process from a vocabulary of only single characters and special symbols, space is inserted between every pair of consecutive characters within each word (keys of the dictionary `tokenFreqs`).\n",
+    "In other words, space is the delimiter between symbols within a word.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "HashMap<String, Integer> rawTokenFreqs = new HashMap<>();\n",
+    "rawTokenFreqs.put(\"fast_\", 4);\n",
+    "rawTokenFreqs.put(\"faster_\", 3);\n",
+    "rawTokenFreqs.put(\"tall_\", 5);\n",
+    "rawTokenFreqs.put(\"taller_\", 4);\n",
+    "\n",
+    "LinkedHashMap<String, Integer> tokenFreqs = new LinkedHashMap<>();\n",
```

Comment:
nit: looks like it could be regular HashMap. The order is changed anyway and doesn't really matter